### PR TITLE
CORE-6690 consistent overrides

### PR DESCRIPTION
**Pull request by JasonSinclair95**
This ticket is incomplete however I recommend we merge what has been done as a-lot has changed. This PR is to make changes so that the Corda workers can query an existing secret for both the db password and username
 
To test the changes the following should be put in your overrides 
```yaml
db:
  cluster:
    host: prereqs-postgresql 
    password:
      valueFrom:
        secretkeyRef:
          name: test-secret
          passwordKey: password 

    user:
      valueFrom: 
        secretkeyRef:
          name: test-secret2
          usernameKey: namename
```
For this deployment i created a secret for the username and one for the password. It was also tested using the same secret and no secret at all.

### CODE REVIEW

**Review by oscardalmau-r3 — DISMISSED** [review]
LGTM

File: charts/corda/values.yaml [hunk 1]
```diff
@@ -60,13 +60,30 @@ db:
...
+      # -- the cluster db username
+      value: user
+      # -- the username secret configuration
+      valueFrom: 
+        # -- the username secret key reference
+        secretkeyRef:
+          # -- the username secret name 
+          name: ""
+          # the username secret key 
+          usernameKey: ""
```
Comment (davidcurrie) [inline charts/corda/values.yaml]:
```suggestion
    user:
      # -- the cluster database user
      value: user
      # the cluster database user secret configuration; used in preference to value if name is set
      valueFrom: 
        # the cluster database user secret key reference
        secretKeyRef:
          # -- the cluster database user secret name 
          name: ""
          # -- the cluster database user secret key 
          key: ""
```

File: charts/corda/values.yaml [hunk 2]
```diff
@@ -60,13 +60,30 @@ db:
...
     # -- the cluster database password (ignored if existingSecret is set, otherwise required)
```
Comment (davidcurrie) [inline charts/corda/values.yaml]:
```suggestion
    # the cluster database password configuration
```

File: charts/corda/values.yaml [hunk 3]
```diff
@@ -60,13 +60,30 @@ db:
     # -- the cluster database port
     port: 5432
     # -- the cluster database user
```
Comment (davidcurrie) [inline charts/corda/values.yaml]:
```suggestion
    # the cluster database user configuration
```

File: charts/corda/values.yaml [hunk 4]
```diff
@@ -60,13 +60,30 @@ db:
...
+      # -- the cluster passowrd
+      value: "" 
+      # -- the password secret configuration
+      valueFrom:
+        # -- the password secret key reference
+        secretkeyRef:
+          # -- the password secret name
+          name: ""
+          # -- the password secret key
+          passwordKey: ""
```
Comment (davidcurrie) [inline charts/corda/values.yaml]:
```suggestion
    password: 
      # -- the cluster database password
      value: "" 
      # the cluster database password secret configuration; used in preference to value if name is set
      valueFrom:
        # the cluster database password secret key reference
        secretKeyRef:
          # -- the cluster database password secret name
          name: ""
          # -- the cluster database password secret key
          key: ""
```

File: charts/corda/values.schema.json [hunk 5]
```diff
@@ -246,13 +246,71 @@
                             ]
                         },
                         "user": {
-                            "type": "string",
-                            "default": "user",
-                            "title": "the cluster database user",
-                            "examples": [
-                                "user"
+                            "type": "object",
```
Comment (davidcurrie) [inline charts/corda/values.schema.json]: schema will need updating to reflect changes proposed in values

File: charts/corda/templates/cluster-db-secret.yaml [hunk 6]
```diff
@@ -1,4 +1,4 @@
-{{- if not .Values.db.cluster.existingSecret }}
+{{- if not .Values.db.cluster.password.valueFrom.secretkeyRef.name }}
 apiVersion: v1
 kind: Secret
```
Comment (davidcurrie) [inline charts/corda/templates/cluster-db-secret.yaml:3]: We should also put the user in this secret (which means the secret needs to be created if either the user or password is not specified via a user-provided secret.

File: charts/corda/templates/_helpers.tpl [hunk 7]
```diff
@@ -370,8 +379,13 @@ Cluster DB name
 {{- end -}}
 
 {{/*
-Cluster DB secret name
+Cluster DB password secret 
 */}}
 {{- define "corda.clusterDbSecretName" -}}
```
Comment (davidcurrie) [inline charts/corda/templates/_helpers.tpl]:
```suggestion
{{- define "corda.clusterDbPasswordEnv" -}}
```

File: charts/corda/templates/_helpers.tpl [hunk 8]
```diff
@@ -356,10 +356,19 @@ Cluster DB port
 {{- end -}}
 
 {{/*
-Cluster DB user
+Cluster DB user secret 
 */}}
 {{- define "corda.clusterDbUser" -}}
```
Comment (davidcurrie) [inline charts/corda/templates/_helpers.tpl]:
```suggestion
{{- define "corda.clusterDbUseEnvr" -}}
```

File: charts/corda/templates/_helpers.tpl [hunk 9]
```diff
@@ -356,10 +356,19 @@ Cluster DB port
...
+{{- if .Values.db.cluster.user.valueFrom.secretkeyRef.name }}
+- name: PGUSER
+  valueFrom:
+    secretKeyRef:
+      name: {{ .Values.db.cluster.user.valueFrom.secretkeyRef.name }}
+      key: {{ .Values.db.cluster.user.valueFrom.secretkeyRef.usernameKey }}
+{{- else }}
+- name: PGUSER
+  value: {{ .Values.db.cluster.user.value }}
+{{- end }}
```
Comment (davidcurrie) [inline charts/corda/templates/_helpers.tpl]: Will need to retrieve from secret if not set (as for password)

File: charts/corda/templates/_helpers.tpl [hunk 10]
```diff
@@ -370,8 +379,13 @@ Cluster DB name
 {{- end -}}
 
 {{/*
-Cluster DB secret name
+Cluster DB password secret
```
Comment (davidcurrie) [inline charts/corda/templates/_helpers.tpl]:
```suggestion
Cluster DB password environment variable 
```

File: charts/corda/templates/_helpers.tpl [hunk 11]
```diff
@@ -356,10 +356,19 @@ Cluster DB port
 {{- end -}}
 
 {{/*
-Cluster DB user
+Cluster DB user secret
```
Comment (davidcurrie) [inline charts/corda/templates/_helpers.tpl]:
```suggestion
Cluster DB user environment variable 
```

File: charts/corda/templates/cluster-db-secret.yaml [hunk 12]
```diff
@@ -10,5 +10,5 @@ metadata:
     {{- include "corda.labels" . | nindent 4 }}
 type: Opaque
 data:
-  password: {{ required "Must specify db.cluster.existingSecret or db.cluster.password" .Values.db.cluster.password | b64enc | quote }}
-{{- end }}
\ No newline at end of file
+  password: {{ required "Must specify db.cluster.password.valueFrom.secretRef.name or db.cluster.password.value" .Values.db.cluster.password.value | b64enc | quote }}
```
Comment (davidcurrie) [inline charts/corda/templates/cluster-db-secret.yaml:17]: the `password` key here needs to use the `key` name specified in the values (or default to `password`)

**Review by davidcurrie — CHANGES_REQUESTED** [review]

File: charts/corda/templates/cluster-db-secret.yaml (hunk 12 above)
Comment (JasonSinclair95) [reply charts/corda/templates/cluster-db-secret.yaml:17]: The default of this value .Values.db.cluster.password.value  is password @davidcurrie

Comment (davidcurrie) [reply charts/corda/templates/cluster-db-secret.yaml:17]: I'm talking about the key here though i.e. the bit before the colon at the start of the line. You're always calling it `password` here but when you retrieve it, you use the value from `secretRef.key`.

File: charts/corda/templates/_helpers.tpl (hunk 9 above)
Comment (JasonSinclair95) [reply charts/corda/templates/_helpers.tpl]: That's what the function is. it looks for the value for the secret if set and if not it gets the default value which is user. Or is this not what you're suggesting @davidcurrie

File: charts/corda/templates/cluster-db-secret.yaml (hunk 6 above)
Comment (JasonSinclair95) [reply charts/corda/templates/cluster-db-secret.yaml:3]: Okay

File: charts/corda/templates/_helpers.tpl (hunk 9 above)
Comment (davidcurrie) [reply charts/corda/templates/_helpers.tpl]: Hopefully it made sense by the time you got to the comment about adding the user to the secret.

File: charts/corda/templates/cluster-db-secret.yaml (hunk 12 above)
Comment (JasonSinclair95) [reply charts/corda/templates/cluster-db-secret.yaml:17]: Could we have a conversation regarding this if statement please

Comment (JasonSinclair95) [reply charts/corda/templates/cluster-db-secret.yaml:17]: The key for the password should always be secret here, as this secret only gets created if SecretRef.name is not set. I don't see how any of the values are retrieved from SecretRef for this secret. @davidcurrie

File: charts/corda/templates/_helpers.tpl [hunk 13]
```diff
@@ -221,6 +221,7 @@ Worker environment variables
 - name: CONSOLE_LOG_FORMAT
   value: {{ .Values.logging.format | quote }}
 - name: CONSOLE_LOG_LEVEL
+<<<<<<< HEAD
```
Comment (yarivus) [inline charts/corda/templates/_helpers.tpl]: bad merge

**Review by yarivus — CHANGES_REQUESTED** [review]

**Review by yarivus — DISMISSED** [review]
LGTM

**Review by oscardalmau-r3 — APPROVED** [review]
LGTM

**Review by davidcurrie — APPROVED** [review]

**Review by ronanbrowne — APPROVED** [review]
LGTM

**Pull request merged**